web/src/message_fetch.js

```diff
@@ -590,83 +512,37 @@ export function initialize(home_view_loaded) {
-        cont: load_more,
+    // Since `all_messages_data` contains continuous message history
+    // with always contains the latest message, it makes sense for
+    // Recent view to display the same data and be in sync.
+    all_messages_data.set_add_messages_callback((messages) => {
+        recent_view_ui.process_messages(messages, all_messages_data);
     });
 
-    if (page_params.is_spectator) {
-        // Since spectators never have old unreads, we can skip the
-        // hacky fetch below for them (which would just waste resources).
-
-        // This optimization requires a bit of duplicated loading
-        // indicator code, here and hiding logic in hide_more.
-        recent_view_ui.show_loading_indicator();
-        return;
-    }
-
-    // In addition to the algorithm above, which is designed to ensure
-    // that we fetch all message history eventually starting with the
-    // first unread message, we also need to ensure that the Recent
-    // Topics page contains the very most recent threads on page load.
-    //
-    // Long term, we'll want to replace this with something that's
-    // more performant (i.e. avoids this unnecessary extra fetch the
-    // results of which are basically discarded) and better represents
-    // more than a few hundred messages' history, but this strategy
-    // allows "Recent Conversations" to always show current data (with gaps)
-    // on page load; the data will be complete once the algorithm
-    // above catches up to present.
-    //
-    // (Users will see a weird artifact where Recent Conversations has a gap
-    // between E.g. 6 days ago and 37 days ago while the catchup
-    // process runs, so this strategy still results in problematic
-    // visual artifacts shortly after page load; just more forgivable
-    // ones).
-    //
-    // We only initialize MessageListData here, since we don't
-    // want update the UI and confuse the functions in MessageList.
-    // Recent view can handle the UI updates itself.
-    const recent_view_message_list_data = new message_list_data.MessageListData({
-        filter: new Filter([{operator: "in", operand: "home"}]),
-        excludes_muted_topics: true,
-    });
     // TODO: Ideally we'd have loading indicators for Recent Conversations
     // at both top and bottom be managed by load_messages, but that
     // likely depends on other reorganizations of the early loading
     // sequence.
     recent_view_ui.show_loading_indicator();
     load_messages({
         anchor: "newest",
-        num_before: consts.recent_view_initial_fetch_size,
-        num_after: 0,
-        msg_list_data: recent_view_message_list_data,
-        is_recent_view_data: true,
-        cont: recent_view_ui.hide_loading_indicator,
+        num_before: consts.initial_backfill_fetch_size,
+        // `num_after` is 1 to help with correctly setting the fetch status
+        // without adding special conditions `all_messages_data`.
+        num_after: 1,
```

Comment:
Hmm. Comment could be simplified:

> We set nonzero num_after so that fetch_status.has_found_newest will be updated correctly.

This might not be free to do, though; the server likely does some useless work to try to fetch such messages. 

Hmm, actually, looking at the server configuration, I think it will return `found_newest=True` even with `num_after` being 0 as long as anchor is set to `"newest"`; maybe verify that end-to-end with the browser inspector, and then maybe we leave it as 0 with a different comment noting that detail of the API.

```
    found_oldest = anchored_to_left or (len(before_rows) < num_before)                                  
    found_newest = anchored_to_right or (len(after_rows) < num_after)                                   
```

Reply:
I added these changes to fix the `found_newest` status not being updated correctly:

```js
    const current_fetch_found_oldest = !has_found_oldest && data.found_oldest;
    const current_fetch_found_newest = !has_found_newest && data.found_newest;

    if (opts.num_before > 0 || current_fetch_found_oldest) {
        msg_list_data.fetch_status.finish_older_batch({
            update_loading_indicator,
            found_oldest: data.found_oldest,
            history_limited: data.history_limited,
        });
        message_feed_top_notices.update_top_of_narrow_notices(opts.msg_list);
    }

    if (opts.num_after > 0 || current_fetch_found_newest) {
        opts.fetch_again = msg_list_data.fetch_status.finish_newer_batch(data.messages, {
            update_loading_indicator,
            found_newest: data.found_newest,
        });
    }
```

I guess we don't need a comment there since it is expected that `anchor: newest` will return `found_newest=True`.

Reply:
Done.